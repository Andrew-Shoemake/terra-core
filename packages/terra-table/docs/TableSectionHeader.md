# Terra Table SectionHeader

The Terra Section Header is built upon a table row and can be placed similarly within a table, provided the number of columns are provided to the SectionHeader.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-table`
  - `yarn add terra-table`

## Usage

```jsx
import React from 'react';
import { SectionHeader } from 'terra-table';

<SectionHeader
  key="unique-react-key"
  isCollapsible
  isCollapsed
  metaData={{ myInfo: 'my-data' }}
  onSelect={this.handleSelection}
  colSpan={3}
/>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
