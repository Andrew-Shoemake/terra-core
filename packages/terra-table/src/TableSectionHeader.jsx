import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import TableUtils from './TableUtils';
import styles from './TableSectionHeader.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Number of columns the section header must span
   */
  colSpan: PropTypes.number.isRequired,
  /**
   * Whether or not the section is collapsed.
   */
  isCollapsed: PropTypes.bool,
  /**
   * Whether or not the section can be collapsed.
   */
  isCollapsible: PropTypes.bool,
  /**
   * Optionally sets the heading level. One of 2`, `3`, `4`, `5`, `6`.
   */
  level: PropTypes.oneOf([2, 3, 4, 5, 6]),
  /**
   * The associated metaData to be provided in the onSelect callback.
   */
  // eslint-disable-next-line react/forbid-prop-types
  metaData: PropTypes.object,
  /**
   * Function callback for when the appropriate click or key action is performed.
   * Callback contains the javascript evnt and prop metadata, e.g. onSelect(event, metaData)
   */
  onSelect: PropTypes.func,
  /**
   * Function callback passthrough for the ref of the section li.
   */
  refCallback: PropTypes.func,
  /**
   * Title text to be placed within the section header.
   */
  title: PropTypes.string,
};

const defaultProps = {
  isCollapsed: false,
  isCollapsible: false,
  level: 2,
  title: '',
};

const TableSectionHeader = ({
  colSpan,
  isCollapsed,
  isCollapsible,
  level,
  metaData,
  onBlur,
  onClick,
  onKeyDown,
  onMouseDown,
  onSelect,
  refCallback,
  title,
  ...customProps
}) => {
  const sectionHeaderClassNames = cx([
    'section-header',
    { 'is-collapsible': isCollapsible },
    customProps.className,
  ]);

  const attrSpread = {};
  const Element = `h${level}`;
  let titleElement = <Element className={cx('title')}>{title}</Element>;
  let accordionIcon;
  if (isCollapsible) {
    accordionIcon = (
      <div className={cx('start')}>
        <span className={cx(['accordion-icon', { 'is-open': !isCollapsed }])} />
      </div>
    );
    titleElement = (
      <div className={cx('fill')}>
        {titleElement}
      </div>
    );

    attrSpread.onClick = TableUtils.wrappedOnClickForItem(onClick, onSelect, metaData);
    attrSpread.onKeyDown = TableUtils.wrappedOnKeyDownForItem(onKeyDown, onSelect, metaData);
    attrSpread.tabIndex = '0';
    attrSpread.role = 'heading';
    attrSpread['aria-expanded'] = !isCollapsed;
    attrSpread['aria-level'] = 1;
    attrSpread['data-row-show-focus'] = 'true';
    attrSpread.onBlur = TableUtils.wrappedEventCallback(onBlur, event => event.currentTarget.setAttribute('data-row-show-focus', 'true'));
    attrSpread.onMouseDown = TableUtils.wrappedEventCallback(onMouseDown, event => event.currentTarget.setAttribute('data-row-show-focus', 'false'));
  }

  return (
    <tr {...customProps} {...attrSpread} className={sectionHeaderClassNames} ref={refCallback}>
      <td colSpan={colSpan}>
        <div className={cx('section-content')}>
          {accordionIcon}
          {titleElement}
        </div>
      </td>
    </tr>
  );
};

TableSectionHeader.propTypes = propTypes;
TableSectionHeader.defaultProps = defaultProps;

export default TableSectionHeader;
