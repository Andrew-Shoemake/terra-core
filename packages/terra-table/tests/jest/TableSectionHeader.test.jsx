import React from 'react';
import TableSectionHeader from '../../src/TableSectionHeader';

// Snapshot test
it('should render a default table section header', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} />,
  );
  expect(section).toMatchSnapshot();
});


it('should render a table section header with isCollpased', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} isCollapsed />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section header with isCollapsible', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} isCollapsible />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section header with level', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} level={2} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section header with metaData', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} metaData={{}} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section header with onSelect', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} onSelect={jest.fn()} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section header with refCallback', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} onSelect={jest.fn()} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section header with title', () => {
  const section = shallow(
    <TableSectionHeader colSpan={1} title="test" />,
  );
  expect(section).toMatchSnapshot();
});
