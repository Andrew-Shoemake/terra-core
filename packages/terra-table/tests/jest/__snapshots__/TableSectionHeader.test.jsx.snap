// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default table section header 1`] = `
<tr
  className="section-header"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <h2
        className="title"
      >
        test
      </h2>
    </div>
  </td>
</tr>
`;

exports[`should render a table section header with isCollapsible 1`] = `
<tr
  aria-expanded={true}
  aria-level={1}
  className="section-header is-collapsible"
  data-row-show-focus="true"
  onBlur={[Function]}
  onMouseDown={[Function]}
  role="heading"
  tabIndex="0"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <div
        className="start"
      >
        <span
          className="accordion-icon is-open"
        />
      </div>
      <div
        className="fill"
      >
        <h2
          className="title"
        >
          test
        </h2>
      </div>
    </div>
  </td>
</tr>
`;

exports[`should render a table section header with isCollpased 1`] = `
<tr
  className="section-header"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <h2
        className="title"
      >
        test
      </h2>
    </div>
  </td>
</tr>
`;

exports[`should render a table section header with level 1`] = `
<tr
  className="section-header"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <h3
        className="title"
      >
        test
      </h3>
    </div>
  </td>
</tr>
`;

exports[`should render a table section header with metaData 1`] = `
<tr
  className="section-header"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <h2
        className="title"
      >
        test
      </h2>
    </div>
  </td>
</tr>
`;

exports[`should render a table section header with onSelect 1`] = `
<tr
  className="section-header"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <h2
        className="title"
      >
        test
      </h2>
    </div>
  </td>
</tr>
`;

exports[`should render a table section header with refCallback 1`] = `
<tr
  className="section-header"
>
  <td
    colSpan={1}
  >
    <div
      className="section-content"
    >
      <h2
        className="title"
      >
        test
      </h2>
    </div>
  </td>
</tr>
`;
